File: sim_1_6_1.rb
# program that generates a random number of pairs of lower bound and upper
# bound, each bound is also a random number up to a maximum value, to be used
# by the programs in 1.6.1.sol.*
require 'optparse'

def parse(args)
  options = {}
  parser = OptionParser.new do |opts|
    opts.banner = "Usage: sim_1_6_1.rb -c CASES -u UPPER"

    options[:cases] = 1
    opts.on("-c", "--number-of CASES", "Number of Test Cases") do |opt|
      options[:cases] = opt.to_i
    end

    options[:upper] = 10
    opts.on("-u", "--upper-bound UPPER", "Upper Bound") do |opt|
      options[:upper] = opt.to_i
    end
  end
  parser.parse!(args)

  return options
end

def generate(options)
  options[:cases].times do
    lower = rand(1..options[:upper]/2)
    upper = rand(lower..options[:upper])
    puts("#{lower} #{upper}")
  end
end

generate(parse(ARGV))

File: sim_1_6_1.c
/* program that generates a random number of pairs of lower bound and upper
 * bound, each bound is also a random number up to a maximum value, to be used
 * by the programs in 1.6.1.sol.*
 */
#include <stdio.h>
#include <time.h>
#include <unistd.h> /* optarg */


void main(int argc, char** argv)
{
    char opt;
    int num_cases = 0;
    int upper_bound = 0;
    char* output_file = NULL;

    while ((opt = getopt(argc, argv, "c:r:o:")) != -1)
    {
        switch (opt)
        {
        case 'c':
            num_cases = atoi(optarg);
            break;
        case 'r':
            upper_bound = atoi(optarg);
            break;
        case 'o':
            output_file = optarg;
            break;
        default:
            printf("Usage: %s -c num_cases -r upper_bound -o output_file\n",
                   argv[0]);
        }
    }

    if (num_cases == 0 || upper_bound == 0 || output_file == NULL)
        printf("Usage: %s -c num_cases -r upper_bound -o output_file\n",
               argv[0]);
    else
    {
        srand(time(NULL));

        int i;
        FILE* file = fopen(output_file, "w");
        for (i = 0; i < num_cases; i++)
        {
            int lower = rand() % (upper_bound-1) + 1;
            int upper = rand() % (upper_bound-lower) + lower + 1;
            fprintf(file, "%d %d\n", lower, upper);
        }

        fclose(file);
    }
}
